server/conn.go

```diff
@@ -990,13 +1005,17 @@ func (cc *clientConn) writeChunks(ctx context.Context, rs ResultSet, binary bool
 				return errors.Trace(err)
 			}
 		}
+		// if cursor exists, checking whether this batch has fetched enough rows.
+		if mysql.HasCursorExistsFlag(serverStatus) && fetchRows >= fetchSize {
```

Comment:
Is it ok to return rows more than `fetchSize`?

Reply:
The client uses cursor for getting results in batches, instead of acquiring the whole result set at once. The size of each batch is controllable to avoid client Out of memory.

The client is still traversing the result set as if there is no cursor. It doesn't care about the fetch size, as long as it does not cause the client OOM.
So I think the current implementation is ok.

Of course, if the fetch size is strictly equal to the value specified by the client, it's a better implementation.
However, I think it is acceptable to meet the essential requirement with minimal modification.

Reply:
I worried about that some of the client has an assumption that the returned result set is no more than fetch size, then it will discard some rows in the result set.

Reply:
If we support strict fetch size, then the change would be too much, then we better define another function without modifying the `writeChunks` in case the change introduces bugs.

Reply:
How about the following solution:

if `fetch size` < `MaxChunkSize`, then `init Chunk capacity` = `fetch size`
if `fetch size` >= `MaxChunkSize`, then `init Chunk capacity` = `MaxChunkSize`

Every time we handle the `COM_STMT_FETCH` command，we fill the chunk with rows only once. 
So the result size must be <= `fetch size`

Reply:
The `init chunk capacity` doesn't matter, the number of rows in a chunk is `MaxChunkSize`.

Reply:
I notice `maxChunkSize` is the property of `baseExecutor`
 
Assuming I add a new method:

`NewChunk(chunkSize)`

It returns a new Chunk with init chunkSize, and setting the current baseExecutor's `maxChunkSize =  chunkSize` .  Doing this will not affect other executors.

how about this way?  Could you please give me some suggestions?

Reply:
I think this feature is not widely used, we can accept this feature only if it doesn't affect the existing logic.

We don't want this feature to add complexity to our normal execution code.

I think the best way is to define another method like `writeChunksWithFetchSize`,  the old `writeChunks` is not touched. 

Common code can be extracted to method though, that is ok.

Reply:
I think @coocood means "TiDB still read data as chunk, but flush output to client use `fetchSize`"? and TiDB hold a partial result buffer when chunk > fetchSize ????

Reply:
@mz1999 PTAL, modify chunk/executor is prefect solution but maybe need too big change.

PS: maybe add some unit test for this modify, current golang are not support server sider cursor integration test seem be not easy, thx~

Reply:
@mz1999 Great!